proposals/0000-slurp.rst

```diff
@@ -0,0 +1,454 @@
+the maintainer of ``bytestring``, Hackage lets you "own" the
+``https://hackage.haskell.org/package/bytestring`` URL, so it's a legit
+place to store all the different versions of ``bytestring``. But as the
+maintainer, you could as well have decided to store the package
+elsewhere instead. Why do we need to force the whole package universe
+into Hackage? Now we don't need to. People can host packages on Hackage
+as before, they can host them directly on Stackage, or indeed on their
+own publicly available corporate server.
+
+Packages can also be hosted *directly on GitHub*.
+Anyone in the community is free to implement a proxy service that maps
+calls to the “package API” above to downloads of GitHub-hosted
+repository snapshots. For that, all a user needs to do is claim a new
+package name on SLURP, once, pointing to the proxy service. Then, the
+proxy service can advertise one package version per Git tag found in the
+repository. Publishing a new (GPG signed!) package version can be done
+with a simple
+
+.. code:: console
+
+    $ git tag -S v1.2.3
+    $ git push --tags origin master
+
+Each package repository (Hackage, Stackage, SomeOtherPackageHost etc) is
+free to define their own policies. Hackage can choose to stick to
+mutable packages as they do now. Stackage can decide that all packages
+uploaded there are always immutable. Hackage can decide to reject
+packages that don't stick to the PVP. Stackage can impose their own
+rules.
+
+6.2 Liberal namespacing
+~~~~~~~~~~~~~~~~~~~~~~~
+
+There have been many suggestions made offline in recent months about
+introducing a hierarchical package namespace for various purposes. One
+such purpose could be to cordon off experimental packages into a separate
+``experimental/`` namespace before they make it into the main namespace.
+Another might be to identify a set of packages specific to
+an as-yet-unreleased version of GHC.
+
+This proposal is compatible with future extensions to the namespace as
+served currently by Hackage. But we do not propose to make any changes
+to this namespace yet. That is a matter for future proposals.
+
+6.3 Minimal changes to Hackage
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Due to the federation, one small change is necessary to
+``hackage-server``, the implementation behind hackage.haskell.org.
+Before accepting a new package, ``hackage-server`` must first
+successfully register the package name on SLURP. If it cannot, then it
+refuses the package upload. This is the only change necessary to
+``hackage-server``.
+
+It would be highly desirable that packages available elsewhere are also
+available to legacy cabal-install versions. To achieve this,
+``hackage-server`` is free to include ``.cabal`` files in the metadata
+bundle that correspond to package versions hosted elsewhere but
+registered via SLURP. Just like Stackage today imports package versions
+available from Hackage.
+
+6.4 No need for extra user accounts
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Comment:
What about the work that went into making the Haskell package update and delivery mechanism not embarrassing insecure? Both Stackage (with signed, S3-hosted indexes) and hackage (with hackage security) have given these things some serious thought. Are we going to lose that again, sacrificed on the altar of the least common denominator?

Reply:
IIUC Stack [also uses `hackage-security` now](https://www.snoyman.com/blog/2017/02/hackage-security-stack).